### PR TITLE
Fix binary verification with modern cmake

### LINKED ISSUE #2101: CMake modifying installed binary causing verification fail
>[!NOTE]
> This does **not** affect binaries built on the CI, only **binaries built locally**

When running `./build.sh` in a local repository, we get a local binary where `./pihole-FTL verify` works as expected. However, when running `/.build.sh install`, any call to `pihole-FTL verify` will actually **fail**.

Furthermore, we observe this in `/var/log/pihole/FTL.log`:
```
2024-10-27 10:10:45.636 CET [3305335M] CRIT: Corrupt binary detected - this may lead to unexpected behaviour!
```

Preliminary investigation results are that **CMake modifies the binary post-build** (i.e., post checksum-generation) which subsequently leads to a hash mismatch and - entirely correctly - emitting the critical error above. This is a new issue after recently upgrading my Pi-hole from Ubuntu 22.04 LTS to 24.04 LTS so it is probably caused by a more recent CMake version being used.

When directly comparing the local and the installed file, we get a few differences:
```
sudo ./build.sh install
hexdump -C cmake/pihole-FTL > local.hex
hexdump -C $(which pihole-FTL) > global.hex
diff local.hex global.hex
```
```
10673,10674c10673,10674
< 00029b00  32 2e 32 39 00 2f 75 73  72 2f 6c 6f 63 61 6c 2f  |2.29./usr/local/|
< 00029b10  6c 69 62 36 34 3a 00 00  00 00 02 00 03 00 04 00  |lib64:..........|
---
> 00029b00  32 2e 32 39 00 00 00 00  00 00 00 00 00 00 00 00  |2.29............|
> 00029b10  00 00 00 00 00 00 00 00  00 00 02 00 03 00 04 00  |................|
398491,398517c398491,398516
< 0061a270  1d 00 00 00 00 00 00 00  65 02 01 00 00 00 00 00  |........e.......|
< 0061a280  0c 00 00 00 00 00 00 00  00 f0 04 00 00 00 00 00  |................|
< 0061a290  0d 00 00 00 00 00 00 00  a4 c5 3a 00 00 00 00 00  |..........:.....|
< 0061a2a0  19 00 00 00 00 00 00 00  50 be 60 00 00 00 00 00  |........P.`.....|
< 0061a2b0  1b 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
< 0061a2c0  1a 00 00 00 00 00 00 00  58 be 60 00 00 00 00 00  |........X.`.....|
< 0061a2d0  1c 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
< 0061a2e0  f5 fe ff 6f 00 00 00 00  b0 03 00 00 00 00 00 00  |...o............|
< 0061a2f0  05 00 00 00 00 00 00 00  a0 98 01 00 00 00 00 00  |................|
< 0061a300  06 00 00 00 00 00 00 00  18 5b 00 00 00 00 00 00  |.........[......|
< 0061a310  0a 00 00 00 00 00 00 00  77 02 01 00 00 00 00 00  |........w.......|
< 0061a320  0b 00 00 00 00 00 00 00  18 00 00 00 00 00 00 00  |................|
< 0061a330  15 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
< 0061a340  03 00 00 00 00 00 00 00  60 b4 61 00 00 00 00 00  |........`.a.....|
< 0061a350  02 00 00 00 00 00 00 00  30 21 00 00 00 00 00 00  |........0!......|
< 0061a360  14 00 00 00 00 00 00 00  07 00 00 00 00 00 00 00  |................|
< 0061a370  17 00 00 00 00 00 00 00  e0 cb 04 00 00 00 00 00  |................|
< 0061a380  07 00 00 00 00 00 00 00  60 b7 02 00 00 00 00 00  |........`.......|
< 0061a390  08 00 00 00 00 00 00 00  80 14 02 00 00 00 00 00  |................|
< 0061a3a0  09 00 00 00 00 00 00 00  18 00 00 00 00 00 00 00  |................|
< 0061a3b0  1e 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
< 0061a3c0  fb ff ff 6f 00 00 00 00  01 00 00 08 00 00 00 00  |...o............|
< 0061a3d0  fe ff ff 6f 00 00 00 00  90 b5 02 00 00 00 00 00  |...o............|
< 0061a3e0  ff ff ff 6f 00 00 00 00  05 00 00 00 00 00 00 00  |...o............|
< 0061a3f0  f0 ff ff 6f 00 00 00 00  18 9b 02 00 00 00 00 00  |...o............|
< 0061a400  f9 ff ff 6f 00 00 00 00  fe 15 00 00 00 00 00 00  |...o............|
< 0061a410  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
---
> 0061a270  0c 00 00 00 00 00 00 00  00 f0 04 00 00 00 00 00  |................|
> 0061a280  0d 00 00 00 00 00 00 00  a4 c5 3a 00 00 00 00 00  |..........:.....|
> 0061a290  19 00 00 00 00 00 00 00  50 be 60 00 00 00 00 00  |........P.`.....|
> 0061a2a0  1b 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
> 0061a2b0  1a 00 00 00 00 00 00 00  58 be 60 00 00 00 00 00  |........X.`.....|
> 0061a2c0  1c 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
> 0061a2d0  f5 fe ff 6f 00 00 00 00  b0 03 00 00 00 00 00 00  |...o............|
> 0061a2e0  05 00 00 00 00 00 00 00  a0 98 01 00 00 00 00 00  |................|
> 0061a2f0  06 00 00 00 00 00 00 00  18 5b 00 00 00 00 00 00  |.........[......|
> 0061a300  0a 00 00 00 00 00 00 00  77 02 01 00 00 00 00 00  |........w.......|
> 0061a310  0b 00 00 00 00 00 00 00  18 00 00 00 00 00 00 00  |................|
> 0061a320  15 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
> 0061a330  03 00 00 00 00 00 00 00  60 b4 61 00 00 00 00 00  |........`.a.....|
> 0061a340  02 00 00 00 00 00 00 00  30 21 00 00 00 00 00 00  |........0!......|
> 0061a350  14 00 00 00 00 00 00 00  07 00 00 00 00 00 00 00  |................|
> 0061a360  17 00 00 00 00 00 00 00  e0 cb 04 00 00 00 00 00  |................|
> 0061a370  07 00 00 00 00 00 00 00  60 b7 02 00 00 00 00 00  |........`.......|
> 0061a380  08 00 00 00 00 00 00 00  80 14 02 00 00 00 00 00  |................|
> 0061a390  09 00 00 00 00 00 00 00  18 00 00 00 00 00 00 00  |................|
> 0061a3a0  1e 00 00 00 00 00 00 00  08 00 00 00 00 00 00 00  |................|
> 0061a3b0  fb ff ff 6f 00 00 00 00  01 00 00 08 00 00 00 00  |...o............|
> 0061a3c0  fe ff ff 6f 00 00 00 00  90 b5 02 00 00 00 00 00  |...o............|
> 0061a3d0  ff ff ff 6f 00 00 00 00  05 00 00 00 00 00 00 00  |...o............|
> 0061a3e0  f0 ff ff 6f 00 00 00 00  18 9b 02 00 00 00 00 00  |...o............|
> 0061a3f0  f9 ff ff 6f 00 00 00 00  fe 15 00 00 00 00 00 00  |...o............|
> 0061a400  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
```

This may be related to the this new output I don't recall having seen before:
```
[...]
Installing pihole-FTL
-- Install configuration: "RelWithDebInfo"
-- Installing: /usr/bin/pihole-FTL
-- Set non-toolchain portion of runtime path of "/usr/bin/pihole-FTL" to ""    <------------
```

There is probably a CMake setting to prevent this binary modification but I'm not aware of it. Another solution would be moving the install step into `./build.sh` instead of doing it with CMake.

```
$ cmake --version
cmake version 3.28.3
```

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Fixes #2101, there is much detail in this issue ticket so I won't duplicate it here. Short story is that CMake modifies the binary's run time path information during the first build and strips this away during install. This obviously modifies the checksum and the verification failed rightfully.

This PR disables modifying run time path information altogether as this is not needed for building FTL. Furthermore, it adds two checks during `./build.sh` to ensure this does not happen again in the future without us noticing immediately and hardening the entire `verify` procedure to skip verification when FTL was somehow built without the checksum actually being appended to the binary.

---

**Related issue or feature (if applicable):** Fixes #2101 

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by DL6ER** [comment]
```
$ ./build.sh install
```
```
-- Compiling dynamically linked executable
-- Building FTL with readline support: YES
-- Building FTL with TLS support: YES
-- Configuring done (0.0s)
-- Generating done (0.0s)
-- Build files have been written to: /home/user/FTL/cmake
[  1%] Built target civetweb
[  3%] Built target cJSON
[  3%] Built target ntp
[  5%] Built target zip
[  6%] Built target api_docs
[  6%] Built target miniz
[  6%] Built target lua_scripts
[  8%] Built target sqlite3
[ 13%] Built target database
[ 20%] Built target tre-regex
[ 26%] Built target config
[ 44%] Built target lua
[ 55%] Built target syscalls
[ 77%] Built target dnsmasq
[ 77%] Built target tomlc99
Building Pi-hole FTL daemon
   - Branch: fix/verify_cmake_install
   - Architecture: x86_64 (compiled locally)
   - Version: v5.25.2-2295-g9664df47
   - Tag: v5.25.2
   - Hash: 9664df47
   - Commit date: 2024-10-27 12:50:36 +0100
[ 79%] Built target tools
[ 79%] Built target ftl_lua
[ 79%] Built target gen_version
[ 81%] Built target webserver
[ 82%] Building C object src/api/CMakeFiles/api.dir/info.c.o
[ 82%] Building C object src/CMakeFiles/core.dir/args.c.o
[ 82%] Building C object src/CMakeFiles/core.dir/log.c.o
[ 91%] Built target api
[100%] Built target core
[100%] Linking C executable ../pihole-FTL
Appending sha256sum to pihole-FTL
SHA256 checksum of pihole-FTL: 96662edb61659cff92a1a1813f3cc9eca3fb9a28fbe942f91ddc78f92e74e732
[100%] Built target pihole-FTL
[✓] Binary integrity check: OK
Installing pihole-FTL
-- Install configuration: "RelWithDebInfo"
-- Installing: /usr/bin/pihole-FTL
[✓] Binary integrity check: OK
```

File: build.sh [hunk 1]
```diff
@@ -82,12 +82,18 @@ fi
...
 # If we are asked to install, we do this here (requires root privileges)
-# Otherwise, we simply copy the binary one level up
+# Otherwise, we simply copy the binary one level down
 if [[ -n "${install}" ]]; then
     echo "Installing pihole-FTL"
     SUDO=$(command -v sudo)
     ${SUDO} cmake --install .
+
+    # Verify checksum again after installation
+    /usr/bin/pihole-FTL verify
```
Comment (yubiuser) [inline build.sh:96]: Is this always the location of the installed binary? Or could this depend on the system the user is running cmake on?

File: src/files.c [hunk 2]
```diff
@@ -785,32 +785,48 @@ bool verify_FTL(bool verbose)
...
 	}
 	fclose(f);
 
+	// Verify checksum mark matches magic string "sha256sum"
+	if(memcmp(checksum_mark, "sha256sum", 9) != 0)
+	{
+		log_warn("Binary integrity check not possible: No checksum mark found");
+		// This is not an error, as the binary may not have a checksum mark
+		// if it was built with a different toolchain
+		return true;
```
Comment (yubiuser) [inline src/files.c:827]: Should we really return `true`here? This is an undefined state, where we can neither verify nor reject the integrity.  We would need a third state. If returning `true` here, the CLI will report binary is Ok and only the log will print the warning.

Comment (DL6ER) [reply src/files.c:827]: Hmm, yes, I agree. Looking into `enum.h` it really doesn't look like we have an aversion against adding `enums` when we need them, so let's add another here:
``` c
enum verify_result {
	VERIFY_OK = 0, // EXIT_SUCCESS
	VERIFY_FAILED, // 1
	VERIFY_ERROR, // 2
	VERIFY_NO_CHECKSUM // 3
} __attribute__ ((packed));
```
which is directly exposed as exit code. As everything `!= 0` is `false` in (ba)sh, this preserves the previous behavior while still exposing more detail to the outside world. I am wondering if we need/should document this somewhere but after looking at `pihole-FTL -h`, this really would go beyond the scope of this "document".

Comment (yubiuser) [reply src/files.c]: I seldom admit it, but more documentation is not necessary I think.

**Review by yubiuser — APPROVED** [review]

**Pull request merged**